### PR TITLE
[FLINK-24371][datastream] Call preCommit on SinkWriter although no committer is available

### PR DESCRIPTION
**Pull request by fapaul**
## What is the purpose of the change

 Before this change, if no committer was after the SinkWriter in the pipeline preCommit was never called. With this change, preCommit is now called as well but the return values are discarded as before.

## Brief change log

  - Call the passed lambda in NoopCommitHandler 

## Verifying this change

- Added a test to ensure that preCommit is called if the Sink is configured without a commiter

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/NoopCommitterHandler.java [hunk 1]
```diff
@@ -34,7 +34,8 @@
 
     @Override
     public List<Object> processCommittables(
-            SupplierWithException<List<Object>, Exception> committableSupplier) {
+            SupplierWithException<List<Object>, Exception> committableSupplier) throws Exception {
+        committableSupplier.get();
```
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/NoopCommitterHandler.java:38]: Why is this needed, what's not working in this way? It feels counter-intuitive to have `preCommit` without `commit`. It seems like I missed a case when I explicitly structured the code to not call `preCommit`.

Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/NoopCommitterHandler.java:38]: If we always want to `preCommit`, then please refactor such that `preCommit` is called on call-site (e.g. remove the `Supplier` and directly pass the result.

**Review by AHeise — COMMENTED** [review]
The motivation is not really clear from the PR/commit message. I have seen the ticket now and it makes a bit more sense but I'd like to think if preCommit is really the best option first.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/NoopCommitterHandler.java (hunk 1 above)
Comment (fapaul) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/NoopCommitterHandler.java:38]: > Why is this needed, what's not working in this way? It feels counter-intuitive to have `preCommit` without `commit`. It seems like I missed a case when I explicitly structured the code to not call `preCommit`.

AFAICT all async sinks which want to implement at-least-once cannot work currently. They are all built on the assumption that during `preCommit` `mailbox.yield()` is called until all outstanding message callbacks are called but they do not have any transaction-like to commit (no committer). I personally experienced the issue when trying to implement the elasticseach sink.

> If we always want to `preCommit`, then please refactor such that `preCommit` is called on call-site (e.g. remove the `Supplier` and directly pass the result.

Sure can do that 👍

Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/NoopCommitterHandler.java]: We should adjust documentation to point out that `prepareCommit = commit` on 1-pc sinks. Except from that, the PR is fine.

File: flink-core/src/main/java/org/apache/flink/api/connector/sink/SinkWriter.java [hunk 2]
```diff
@@ -63,6 +63,9 @@ default void writeWatermark(Watermark watermark) throws IOException, Interrupted
      *
      * <p>This will be called before we checkpoint the Writer's state in Streaming execution mode.
      *
+     * <p>In case the sink implements a 1-phase commit protocol, this method should trigger the
```
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-core/src/main/java/org/apache/flink/api/connector/sink/SinkWriter.java:66]: I wonder if this is really self-explanatory. How about

```suggestion
<p>In case the sink has no explicit committer, this method is still called to allow the writer to implement a 1-phase commit protocol. 
```

[the remark this answers is not in the document]
Comment (AHeise) [reply flink-core/src/main/java/org/apache/flink/api/connector/sink/SinkWriter.java:66]: Sorry for being nitty, but I'd like to have the interfaces in good shape.

[the remark this answers is not in the document]
Comment (fapaul) [reply flink-core/src/main/java/org/apache/flink/api/connector/sink/SinkWriter.java:66]: No worries, your suggestion definitely provides more information about the committer writer relation.
